### PR TITLE
Send telemetry over WebSockets

### PR DESCRIPTION
**Pull request by ZachGarcia42**

### CODE REVIEW

**Comment by ZachGarcia42** [comment]
* **#95** <a href="https://app.graphite.dev/github/pr/cornellrocketryteam/Ground-Software/95?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ZachGarcia42 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: telemetry-proxy/telemetry_proxy.go [hunk 1]
```diff
@@ -42,7 +56,7 @@ func main() {
 
 	// Set up a connection to the grpc server
 	address := os.Getenv("FILL_HOSTNAME") + ":50051"
-	conn, err := grpc.Dial(address, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	conn, err := grpc.NewClient(address, grpc.WithTransportCredentials(insecure.NewCredentials()))
```
[the remark this answers is not in the document]
Comment (ZachGarcia42) [reply telemetry-proxy/telemetry_proxy.go:56]: Dial is deprecated and NewClient is the replacement

**Review by ZachGarcia42 — COMMENTED** [review]
Tested with 10 WebSocket clients at once. All received every packet

File: telemetry-proxy/telemetry_proxy.go [hunk 2]
```diff
@@ -30,7 +31,17 @@ var upgrader = websocket.Upgrader{
 	},
 }
 
+// Create a channel to send Protobuf telemetry messages to the WebSocket handler
+var telemetryChannel = make(chan *pb.Telemetry)
+
+// Map to store active WebSocket connections
+var connections = make(map[*websocket.Conn]bool)
```
Comment (maxslarsson) [inline telemetry-proxy/telemetry_proxy.go:38]: It seems like the map is never set to false. It is either inserted with the value true or removed from the map. Can this be an array instead?

File: telemetry-proxy/telemetry_proxy.go [hunk 3]
```diff
@@ -120,4 +177,6 @@ func websocketHandler(w http.ResponseWriter, r *http.Request) {
 			break
 		}
 	}
+	// Remove the connection from the map when it closes
+	defer delete(connections, conn)
```
Comment (maxslarsson) [inline telemetry-proxy/telemetry_proxy.go:181]: Also, shouldn't this defer be moved up to line 164 (where it is inserted into the connections map)?

**Review by maxslarsson — CHANGES_REQUESTED** [review]

File: telemetry-proxy/telemetry_proxy.go (hunk 2 above)
Comment (ZachGarcia42) [reply telemetry-proxy/telemetry_proxy.go:38]: Arrays in Go are a fixed size I believe, so it's better to use a Map. It also makes looking up by connection easier (we don't have to keep track of indices)

File: telemetry-proxy/telemetry_proxy.go (hunk 3 above)
Comment (ZachGarcia42) [reply telemetry-proxy/telemetry_proxy.go:181]: Good point. Moved there

**Review by maxslarsson — APPROVED** [review]

**Comment by ZachGarcia42** [comment]
### Merge activity

* **Oct 6, 12:18 PM EDT**: @ZachGarcia42 merged this pull request with [Graphite](https://app.graphite.dev/github/pr/cornellrocketryteam/Ground-Software/95).

**Pull request merged**